File: iosApp/iosApp/SwiftUIHelper.swift
import ComposeApp
import SwiftUI

class TheFactory: SwiftUIFactory {
    let controller = MapViewWrapperViewController()
    
    func updateCurrentLatLng(currentLatLng: MapLatLng) { controller.updateCurrentLatLng(currentLatLng) }
    func updateSelectedLatLng(selectedLatLng: MapLatLng) { controller.updateSelectedLatLng(selectedLatLng) }
    func updateMarkers(markers: [KotlinTriple<KotlinInt, KotlinDouble, KotlinDouble>]) { controller.updateMarkers(markers) }
    func updateRoutes(routes: [MapLatLng]) { controller.updateRoutes(routes) }
    func updateCandidates(candidates: [Candidate]) { controller.updateCandidates(candidates) }
    func updateSelectedIndex(selectedIndex: Int32) { controller.updateSelectedIndex(selectedIndex) }
    
    func makeController(
        onSelectLatLng: @escaping (MapLatLng) -> Void,
        onMapClick: @escaping (MapPointF, MapLatLng) -> Void
    ) -> UIViewController {
        controller.mapEnv.onSelectLatLng = onSelectLatLng
        controller.mapEnv.onMapClick = onMapClick
        
        return controller
    }
}

File: composeApp/src/iosMain/kotlin/zone/ien/map/utils/SwiftUIHelper.kt
package zone.ien.map.utils

import androidx.compose.runtime.Composable
import androidx.compose.runtime.MutableState
import androidx.compose.ui.Modifier
import androidx.compose.ui.interop.UIKitViewController
import kotlinx.cinterop.ExperimentalForeignApi
import platform.UIKit.UIViewController
import zone.ien.map.data.Candidate
import zone.ien.map.data.RouteResult

interface SwiftUIFactory {

    companion object {
        var shared: SwiftUIFactory? = null
    }

    fun makeController(
        onSelectLatLng: (MapLatLng) -> Unit,
        onMapClick: (MapPointF, MapLatLng) -> Unit
    ) : UIViewController

    fun updateCurrentLatLng(
        currentLatLng: MapLatLng
    )

    fun updateSelectedLatLng(
        selectedLatLng: MapLatLng
    )

    fun updateMarkers(
        markers: List<Triple<Int, Double, Double>>
    )

    fun updateRoutes(
        routes: List<MapLatLng>
    )

    fun updateCandidates(
        candidates: List<Candidate>
    )

    fun updateSelectedIndex(
        selectedIndex: Int
    )
}